integrations/operator/controllers/reconcilers/generic.go

```diff
@@ -0,0 +1,290 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+// Resource is any Teleport Resource the controller manages.
+type Resource any
+
+// Adapter is an empty struct implementing helper functions for the reconciler
+// to extract information from the Resource. This avoids having to implement the
+// same interface on all resources. This became an issue as new resources are
+// not implementing the types.Resource interface anymore.
+type Adapter[T Resource] interface {
+	GetResourceName(T) string
+	GetResourceRevision(T) string
+	GetResourceOrigin(T) string
+	SetResourceRevision(T, string)
+	SetResourceLabels(T, map[string]string)
+}
+
+// KubernetesCR is a Kubernetes CustomResource representing a Teleport Resource.
+type KubernetesCR[T Resource] interface {
+	kclient.Object
+	ToTeleport() T
+	StatusConditions() *[]v1.Condition
+}
+
+// resourceClient is a CRUD client for a specific Teleport Resource.
+// Implementing this interface allows to be reconciled by the resourceReconciler
+// instead of writing a new specific reconciliation loop.
+// resourceClient implementations can optionally implement the resourceMutator
+// and existingResourceMutator interfaces.
+type resourceClient[T Resource] interface {
+	Get(context.Context, string) (T, error)
+	Create(context.Context, T) error
+	Update(context.Context, T) error
+	Delete(context.Context, string) error
+}
+
+// resourceMutator can be implemented by resourceClients
+// to edit a Resource before its creation/update.
+type resourceMutator[T Resource] interface {
+	Mutate(new T)
+}
+
+// existingResourceMutator can be implemented by TeleportResourceClients
+// to edit a Resource before its update based on the existing one.
+type existingResourceMutator[T Resource] interface {
+	MutateExisting(new, existing T)
+}
+
+// resourceReconciler is a Teleport generic reconciler.
+type resourceReconciler[T any, K KubernetesCR[T]] struct {
+	ResourceBaseReconciler
+	resourceClient resourceClient[T]
+	gvk            schema.GroupVersionKind
+	adapter        Adapter[T]
+}
+
+// Upsert is the resourceReconciler of the ResourceBaseReconciler UpsertExternal
+// It contains the logic to check if the Resource already exists, if it is owned by the operator and what
+// to do to reconcile the Teleport Resource based on the Kubernetes one.
+func (r resourceReconciler[T, K]) Upsert(ctx context.Context, obj kclient.Object) error {
+	debugLog := ctrllog.FromContext(ctx).V(1)
+	u, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("failed to convert Object into Resource object: %T", obj)
+	}
+	k8sResource := newKubeResource[K]()
+	debugLog.Info(fmt.Sprintf("Converting resource from unstructured to %T", k8sResource))
```

Comment:
Nit: it'd be preferable to use structured logging here. The reason you've had to use Sprintf is because they are trying to encourage people to switch to structured logging.